src/entropy.rs

```diff
@@ -0,0 +1,339 @@
+    ///
+    /// ## Remarks
+    ///
+    /// The Kullback-Leibler divergence is a measure used in [Information Theory]
+    /// to describe the relationship between two probability distribution: it only make sense
+    /// when each array sums to 1 with entries between 0 and 1 (extremes included).
+    ///
+    /// By definition, *pᵢ ln(qᵢ/pᵢ)* is set to 0 if *pᵢ* is 0.
+    ///
+    /// [Kullback-Leibler divergence]: https://en.wikipedia.org/wiki/Kullback%E2%80%93Leibler_divergence
+    /// [Information Theory]: https://en.wikipedia.org/wiki/Information_theory
+    fn kl_divergence(&self, q: &Self) -> Option<A>
+        where
+            A: Float;
+
+    /// Computes the [cross entropy] *H(p,q)* between two arrays,
+    /// where `self`=*p*.
+    ///
+    /// The cross entropy is defined as:
+    ///
+    /// ```text
+    ///            n
+    /// H(p,q) = - ∑ pᵢ ln(qᵢ)
+    ///           i=1
+    /// ```
+    ///
+    /// If the arrays are empty or their lengths are not equal, `None` is returned.
+    ///
+    /// **Panics** if any element in *q* is negative and taking the logarithm of a negative number
+    /// is a panic cause for `A`.
+    ///
+    /// ## Remarks
+    ///
+    /// The cross entropy is a measure used in [Information Theory]
+    /// to describe the relationship between two probability distribution: it only make sense
+    /// when each array sums to 1 with entries between 0 and 1 (extremes included).
+    ///
+    /// The cross entropy is often used as an objective/loss function in
+    /// [optimization problems], including [machine learning].
+    ///
+    /// By definition, *pᵢ ln(qᵢ)* is set to 0 if *pᵢ* is 0.
+    ///
+    /// [cross entropy]: https://en.wikipedia.org/wiki/Cross-entropy
+    /// [Information Theory]: https://en.wikipedia.org/wiki/Information_theory
+    /// [optimization problems]: https://en.wikipedia.org/wiki/Cross-entropy_method
+    /// [machine learning]: https://en.wikipedia.org/wiki/Cross_entropy#Cross-entropy_error_function_and_logistic_regression
+    fn cross_entropy(&self, q: &Self) -> Option<A>
+    where
+        A: Float;
+}
+
+
+impl<A, S, D> EntropyExt<A, S, D> for ArrayBase<S, D>
+    where
+        S: Data<Elem = A>,
+        D: Dimension,
+{
+    fn entropy(&self) -> Option<A>
+        where
+            A: Float
+    {
+        if self.len() == 0 {
+            None
+        } else {
+            let entropy = self.map(
+                |x| {
+                    if *x == A::zero() {
+                        A::zero()
+                    } else {
+                        *x * x.ln()
+                    }
+                }
+            ).sum();
+            Some(-entropy)
+        }
+    }
+
+    fn kl_divergence(&self, q: &Self) -> Option<A>
+        where
+            A: Float
+    {
+        if (self.len() == 0) | (self.len() != q.len()) {
+            None
+        } else {
+            let kl_divergence: A = self.iter().zip(q.iter()).map(
+                |(p, q)| {
```

Comment:
Just a style thing -- when I have a closure that takes references to a type that implements `Copy`, I usually write the closure like this:

```rust
|(&p, &q)| {
    if p == A::zero() {
        p
    } else {
        p * (q / p).ln()
    }
}
```

so that I don't have to explicitly dereference the parameters in the body every time I use them.